Content.Server/Magic/MagicSystem.cs

```diff
@@ -69,18 +70,36 @@ private void OnDoAfter(EntityUid uid, SpellbookComponent component, DoAfterEvent
         if (args.Handled || args.Cancelled)
             return;
 
-        _actionsSystem.AddActions(args.Args.User, component.Spells, component.LearnPermanently ? null : uid);
         args.Handled = true;
+        if (!component.LearnPermanently)
+        {
+            _actionsSystem.GrantActions(args.Args.User, component.Spells, uid);
+            return;
+        }
+
+        foreach (var (id, charges) in component.SpellActions)
+        {
+            EntityUid? actionId = null;
+            if (_actionsSystem.AddAction(uid, ref actionId, id))
+                _actionsSystem.SetCharges(actionId, charges < 0 ? null : charges);
+        }
+
+        component.SpellActions.Clear();
     }
 
-    private void OnInit(EntityUid uid, SpellbookComponent component, ComponentInit args)
+    private void OnInit(EntityUid uid, SpellbookComponent component, MapInitEvent args)
     {
-        //Negative charges means the spell can be used without it running out.
+        if (!component.LearnPermanently)
+            return;
+
         foreach (var (id, charges) in component.SpellActions)
         {
-            var spell = Spawn(id);
+            var spell = _actionContainer.AddAction(uid, id);
+            if (spell == null)
+                continue;
+
             _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
-            component.Spells.Add(spell);
+            component.Spells.Add(spell.Value);
```

Comment:
I'm in the midst of refactoring magic so spellbooks are going to be changing drastically. To prevent any major conflicts please only include changes required for action containers and revert any changes done to spellbook usage.